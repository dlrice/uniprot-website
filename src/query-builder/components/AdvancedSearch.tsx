import React, { FC, FormEvent, useState, useEffect } from 'react';
import { useHistory, useRouteMatch, generatePath } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import qs from 'query-string';
import { PageIntro } from 'franklin-sites';

import ClauseList from './ClauseList';

import useDataApi from '../../shared/hooks/useDataApi';

import { createEmptyClause, createPreSelectedClauses } from '../utils/clause';
import { stringify, parse } from '../utils/queryStringProcessor';

import { addMessage } from '../../messages/state/messagesActions';

import apiUrls from '../../shared/config/apiUrls';
import { Namespace, NamespaceLabels } from '../../shared/types/namespaces';
import { LocationToPath, Location } from '../../app/config/urls';

import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import { Clause, SearchTermType } from '../types/searchTypes';

import '../../uniprotkb/components/search/styles/search-container.scss';
import './styles/advanced-search.scss';

const flatten = (searchTermData: SearchTermType[]): SearchTermType[] => {
  return searchTermData.flatMap((searchTermDatum: SearchTermType) => {
    if (searchTermDatum.siblings) {
      return flatten(searchTermDatum.siblings);
    }
    if (searchTermDatum.items) {
      return flatten(searchTermDatum.items);
    }
    return searchTermDatum;
  });
};

const AdvancedSearch: FC = () => {
  const history = useHistory();
  const match = useRouteMatch<{ namespace?: Namespace }>(
    LocationToPath[Location.QueryBuilder]
  );
  const dispatch = useDispatch();

  // To be replaced by getting it from url
  const [clauses, setClauses] = useState<Clause[]>([]);

  const namespace = match?.params?.namespace;

  const { data: searchTermsData } = useDataApi<SearchTermType[]>(
    namespace && apiUrls.advancedSearchTerms(namespace)
  );

  // if URL doesn't finish with a namespace redirect to "uniprotkb" by default
  useEffect(() => {
    if (!namespace) {
      history.replace(
        history.createHref({
          ...history.location,
          pathname: generatePath(LocationToPath[Location.QueryBuilder], {
            namespace: Namespace.uniprotkb,
          }),
        })
      );
    }
  }, [history, namespace]);

  useEffect(() => {
    if (!searchTermsData) {
      return;
    }
    const query = qs.parse(history.location.search, { decode: true })?.query;
    const parsedQuery =
      query && !Array.isArray(query) ? parse(query) : undefined;

    // flatten all the endpoint-described clauses to be able to to look-up
    const flattened = flatten(searchTermsData);
    // for each parsed clause, try to find the corresponding endpoint-described
    // clause to merge its 'searchTerm' field
    const validatedQuery: Clause[] = [];
    for (const clause of parsedQuery || []) {
      const found = flattened.find(
        (item) => item.term === clause.searchTerm.term
      );
      // if it exists, assign it 'searchTerm'
      if (found) {
        validatedQuery.push({ ...clause, searchTerm: found });
      } else {
        dispatch(
          addMessage({
            id: clause.searchTerm.term,
            content: `"${clause.searchTerm.term}" is not a valid query term for ${namespace}`,
            format: MessageFormat.POP_UP,
            level: MessageLevel.FAILURE,
          })
        );
      }
    }

    setClauses((clauses) => {
      if (clauses.length) {
        return clauses;
      }
      if (validatedQuery.length) {
        return validatedQuery;
      }
      return createPreSelectedClauses();
    });
  }, [dispatch, history.location.search, namespace, searchTermsData]);

  if (!searchTermsData || !namespace) {
    return null;
  }

  const addClause = () => {
    setClauses((clauses) => [...clauses, createEmptyClause()]);
  };

  const removeClause = (clauseId: string) => {
    setClauses((clauses) => {
      if (clauses.length === 1) {
        return [createEmptyClause()];
      }
      return clauses.filter((clause) => clause.id !== clauseId);
    });
  };

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault();
    const queryString = stringify(clauses);
    history.push({
      pathname: `/${namespace}`,
      search: queryString && `query=${queryString}`,
    });
  };

  return (
    <>
      <PageIntro title="Advanced search" />

      <form
        className="advanced-search"
        onSubmit={handleSubmit}
        data-testid="advanced-search-form"
      >
        <fieldset>
          <label htmlFor="namespace-select">
            Searching in
            <select
              id="namespace-select"
              onChange={(e) => {
                history.replace(
                  history.createHref({
                    pathname: generatePath(
                      LocationToPath[Location.QueryBuilder],
                      { namespace: e.target.value }
                    ),
                  })
                );
              }}
              value={namespace}
            >
              {Object.keys(NamespaceLabels).map((key) => (
                <option value={key} key={key}>
                  {NamespaceLabels[key as Namespace]}
                </option>
              ))}
            </select>
          </label>
        </fieldset>
        <fieldset>
          <ClauseList
            removeClause={removeClause}
            clauses={clauses}
            setClauses={setClauses}
            searchTerms={searchTermsData}
          />
        </fieldset>
        <div className="advanced-search__actions">
          <button
            type="button"
            id="add-field"
            className="button tertiary"
            data-testid="advanced-search-add-field"
            onClick={addClause}
          >
            Add Field
          </button>
          <button type="submit" id="submit-query" className="button">
            Search
          </button>
        </div>
      </form>
    </>
  );
};

export default AdvancedSearch;
