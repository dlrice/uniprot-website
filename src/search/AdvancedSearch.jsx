import React from 'react';
import ClauseList from './ClauseList';

const AdvancedSearch = (props) => {
  const { handleAdvancedSubmitClick, namespace, dispatchAddClause } = props;
  return (
    <div className="advanced-search">
      <div>
        <label htmlFor="namespace-select">
          Searching in
          <select id="namespace-select">
            <option>{namespace}</option>
          </select>
        </label>
      </div>
      <ClauseList {...props} />
      <button type="button" id="add-field" className="button" onClick={dispatchAddClause}>
        Add Field
      </button>
      <div>
        <button
          type="button"
          id="submit-query"
          className="button"
          onClick={handleAdvancedSubmitClick}
        >
          Search
        </button>
      </div>
    </div>
  );
};

export default AdvancedSearch;
