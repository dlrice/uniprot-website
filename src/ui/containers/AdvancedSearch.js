import { connect } from 'react-redux';
import {
  selectField,
  updateInputValue,
  updateEvidence,
  updateRangeValue,
  updateLogicOperator,
  submitQuery,
  addClause,
} from '../actions';
import AdvancedSearchForm from '../components/AdvancedSearchForm';

const mapStateToProps = state => ({
  clauses: state.query.clauses,
  searchTerms: state.query.searchTerms,
  namespace: state.query.namespace,
});

const mapDispatchToProps = dispatch => ({
  handleFieldSelect: (clauseId, field) => dispatch(selectField(clauseId, field)),
  handleInputChange: (clauseId, fieldId, value) => dispatch(updateInputValue(clauseId, fieldId, value)),
  handleEvidenceChange: (clauseId, fieldId, value) => dispatch(updateEvidence(clauseId, fieldId, value)),
  handleRangeInputChange: (clauseId, fieldId, value, from) => dispatch(updateRangeValue(clauseId, fieldId, value, from)),
  handleLogicChange: (clauseId, fieldId, value) => dispatch(updateLogicOperator(clauseId, fieldId, value)),
  handleRemoveClause: clauseId => dispatch(selectField(clauseId)),
  submitQuery: () => dispatch(submitQuery()),
  addClause: () => dispatch(addClause()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdvancedSearchForm);
