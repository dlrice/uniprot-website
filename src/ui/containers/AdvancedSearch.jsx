import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  selectField,
  updateInputValue,
  updateEvidence,
  updateRangeValue,
  updateLogicOperator,
  submitQuery,
  addClause,
  removeClause,
  fetchSearchTerms,
  fetchEvidencesIfNeeded,
} from '../actions';
import ClauseList from '../components/ClauseList';


class AdvancedSearch extends Component {
  componentDidMount() {
    const {
      dispatchFetchSearchTerms,
      dispatchfetchEvidencesIfNeeded,
    } = this.props;
    dispatchFetchSearchTerms();
    dispatchfetchEvidencesIfNeeded('go');
    dispatchfetchEvidencesIfNeeded('annotation');
  }

  render() {
    const { namespace, dispatchAddClause, dispatchSubmitQuery } = this.props;
    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
          <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        <ClauseList
          {...this.props}
        />
        <hr />
        <div>
          <button type="button" id="add-field" className="button" onClick={dispatchAddClause}>
            Add Field
          </button>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={dispatchSubmitQuery}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  clauses: state.query.clauses,
  searchTerms: state.query.searchTerms.data,
  namespace: state.query.namespace,
  evidences: state.query.evidences,
});

const mapDispatchToProps = dispatch => ({
  handleFieldSelect: (clauseId, field) => dispatch(selectField(clauseId, field)),
  handleInputChange: (clauseId, value) => dispatch(updateInputValue(clauseId, value)),
  handleEvidenceChange: (clauseId, value) => dispatch(updateEvidence(clauseId, value)),
  handleRangeInputChange: (clauseId, value, from) => dispatch(updateRangeValue(clauseId, value, from)),
  handleLogicChange: (clauseId, value) => dispatch(updateLogicOperator(clauseId, value)),
  handleRemoveClause: clauseId => dispatch(removeClause(clauseId)),
  dispatchSubmitQuery: () => dispatch(submitQuery()),
  dispatchAddClause: () => dispatch(addClause()),
  dispatchfetchEvidencesIfNeeded: evidencesType => dispatch(fetchEvidencesIfNeeded(evidencesType)),
  dispatchFetchSearchTerms: () => dispatch(fetchSearchTerms()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdvancedSearch);
