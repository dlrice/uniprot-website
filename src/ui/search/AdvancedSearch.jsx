// @flow
import React, { Component } from 'react';
import { v1 } from 'uuid';
import AdvancedSearchField from './AdvancedSearchField';
import withData from '../hoc/withData';
import appendUniqueId from '../hoc/prepareData';
import apiUrls from '../apiUrls';

import type { Field } from './AdvancedSearchField';

type Props = {
  data: [],
};

type State = {
  namespace: string,
  fields: Array<Field>,
};

class AdvancedSearch extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      namespace: 'UniProtKB',
      fields: this.initFields(),
    };
  }

  createField = (): Field => ({
    id: v1(),
    logic: 'AND',
    selectedNode: {
      label: 'Any',
      term: 'All',
      example: 'a4_human, P05067, cdc7 human',
      itemType: 'single',
      dataType: 'string',
    },
    queryInput: {},
  });

  initFields = (): Array<Field> => [...Array(4)].map(() => this.createField());

  updateField = (field: Field) => {
    const { fields } = this.state;
    let match: void | Field = fields.find(f => f.id === field.id);
    if (match) {
      match = field;
      this.setState({ fields });
    }
  };

  createQueryString = (): string => {
    const { fields } = this.state;
    return fields.reduce((queryAccumulator: string, field: Field) => {
      let query = '';
      if (field.queryInput.rangeFrom || field.queryInput.rangeTo) {
        query = `${query}(${field.selectedNode.term}:[${
          field.queryInput.rangeFrom ? field.queryInput.rangeFrom : ''
          }-${field.queryInput.rangeTo ? field.queryInput.rangeTo : ''}])`;
      }
      if (field.queryInput.stringValue && field.queryInput.stringValue !== '') {
        query = `${query}(${field.selectedNode.term}:${
          field.queryInput.stringValue ? field.queryInput.stringValue : ''
          })`;
      }
      if (field.queryInput.evidenceValue && field.queryInput.evidenceValue !== '') {
        query = `${query}AND(${field.selectedNode.term}:${field.queryInput.evidenceValue})`;
      }
      return `${queryAccumulator}${
        queryAccumulator.length > 0 && query.length > 0 ? field.logic : ''
        }${query}`;
    }, '');
  };

  addField() {
    const { fields } = this.state;
    fields.push(this.createField());
    this.setState({ fields });
  }

  removeField(id: string) {
    const { fields } = this.state;
    const updatedFields = fields.filter(field => field.id !== id);
    this.setState({ fields: updatedFields });
  }

  submitQuery() {
    console.log(this.createQueryString());
  }

  render() {
    const { namespace, fields } = this.state;
    const { data } = this.props;
    if (!data) {
      return null;
    }
    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
            <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        {fields.map(field => (
          <div key={`field_${field.id}`} className="advanced-search__field">
            <AdvancedSearchField data={data} field={field} updateField={this.updateField} />
            <button
              type="button"
              className="button-remove"
              onClick={() => this.removeField(field.id)}
            >
              Remove
            </button>
          </div>
        ))}
        <hr />
        <div>
          <button type="button" id="add-field" className="button" onClick={() => this.addField()}>
            Add Field
          </button>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={() => this.submitQuery()}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

const url = () => apiUrls.advanced_search_terms;
export default withData(url)(AdvancedSearch);
