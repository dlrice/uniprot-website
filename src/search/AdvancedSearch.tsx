import React, { Component } from 'react';
import ClauseList from './ClauseList';
import { unpackQueryUrl } from '../utils/apiUrls';
import { Namespace } from './types/searchTypes';
import './styles/AdvancedSearch.scss';

type AdvancedSearchProps = {
  handleAdvancedSubmitClick: () => void;
  namespace: Namespace;
  dispatchAddClause: () => void;
};
class AdvancedSearch extends Component<AdvancedSearchProps> {
  componentDidMount() {
    const {
      searchTerms,
      dispatchUpdateClauses,
      dispatchfetchEvidencesIfNeeded,
      dispatchFetchSearchTermsIfNeeded,
      queryString,
    } = this.props;
    dispatchfetchEvidencesIfNeeded('go');
    dispatchfetchEvidencesIfNeeded('annotation');
    dispatchFetchSearchTermsIfNeeded();

    if (searchTerms.length && queryString) {
      dispatchUpdateClauses(unpackQueryUrl(queryString, searchTerms));
    }
  }

  componentDidUpdate(prevProps) {
    const { queryString: prevQueryString, searchTerms: prevSearchTerms } = prevProps;
    const { queryString, searchTerms, dispatchUpdateClauses } = this.props;
    if (
      searchTerms.length
      && queryString
      && (queryString !== prevQueryString || searchTerms !== prevSearchTerms)
    ) {
      dispatchUpdateClauses(unpackQueryUrl(queryString, searchTerms));
    }
  }

  render() {
    const { handleAdvancedSubmitClick, namespace, dispatchAddClause } = this.props;
    return (
      <div className="advanced-search">
        <div>
          <label htmlFor="namespace-select">
            Searching in
            <select id="namespace-select">
              <option>{namespace}</option>
            </select>
          </label>
        </div>
        <ClauseList {...this.props} />
        <button type="button" id="add-field" onClick={dispatchAddClause}>
          Add Field
        </button>
        <div>
          <button
            type="button"
            id="submit-query"
            className="button"
            onClick={handleAdvancedSubmitClick}
          >
            Search
          </button>
        </div>
      </div>
    );
  }
}

export default AdvancedSearch;
