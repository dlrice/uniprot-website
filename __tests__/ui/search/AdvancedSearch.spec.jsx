import React from 'react';
import renderer from 'react-test-renderer';

import AdvancedSearch from '../../../src/ui/search/AdvancedSearch';

describe('AdvancedSearch component', () => {
  test('should render', () => {
    const component = renderer.create(<AdvancedSearch />).toJSON();

    expect(component).toMatchSnapshot();
  });
});
