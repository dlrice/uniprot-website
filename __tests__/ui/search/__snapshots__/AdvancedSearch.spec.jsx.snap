// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSearch component should render 1`] = `
<div
  className="advanced-search"
>
  <div>
    <label
      htmlFor="namespace-select"
    >
      Searching in
      <select
        id="namespace-select"
      >
        <option>
          UniProtKB
        </option>
      </select>
    </label>
  </div>
  <div
    className="advanced-search__field"
  >
    <select
      className="advanced-search__logic"
      onChange={[Function]}
      value="AND"
    >
      <option
        value="AND"
      >
        AND
      </option>
      <option
        value="OR"
      >
        OR
      </option>
      <option
        value="NOT"
      >
        NOT
      </option>
    </select>
    <div
      className="dropdown-container"
    >
      <button
        className="button dropdown"
        onClick={[Function]}
        type="button"
      >
        Select
      </button>
      <div
        className="dropdown-menu"
      >
        <div
          className="dropdown-container"
        >
          <form
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Search for field"
              type="text"
              value=""
            />
          </form>
          <div
            className="autocomplete-menu"
          >
            <div
              className="dropdown-menu__panel"
            />
          </div>
        </div>
        <div
          className="dropdown-menu__panel"
        >
          <ul
            className=""
          />
        </div>
      </div>
    </div>
    <div
      className="advanced-search__inputs"
    >
      <label
        htmlFor="input_All"
      >
        Any
        <input
          id="input_All"
          onChange={[Function]}
          placeholder="a4_human, P05067, cdc7 human"
          type="text"
        />
      </label>
    </div>
    <button
      className="button-remove"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
  <div
    className="advanced-search__field"
  >
    <select
      className="advanced-search__logic"
      onChange={[Function]}
      value="AND"
    >
      <option
        value="AND"
      >
        AND
      </option>
      <option
        value="OR"
      >
        OR
      </option>
      <option
        value="NOT"
      >
        NOT
      </option>
    </select>
    <div
      className="dropdown-container"
    >
      <button
        className="button dropdown"
        onClick={[Function]}
        type="button"
      >
        Select
      </button>
      <div
        className="dropdown-menu"
      >
        <div
          className="dropdown-container"
        >
          <form
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Search for field"
              type="text"
              value=""
            />
          </form>
          <div
            className="autocomplete-menu"
          >
            <div
              className="dropdown-menu__panel"
            />
          </div>
        </div>
        <div
          className="dropdown-menu__panel"
        >
          <ul
            className=""
          />
        </div>
      </div>
    </div>
    <div
      className="advanced-search__inputs"
    >
      <label
        htmlFor="input_All"
      >
        Any
        <input
          id="input_All"
          onChange={[Function]}
          placeholder="a4_human, P05067, cdc7 human"
          type="text"
        />
      </label>
    </div>
    <button
      className="button-remove"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
  <div
    className="advanced-search__field"
  >
    <select
      className="advanced-search__logic"
      onChange={[Function]}
      value="AND"
    >
      <option
        value="AND"
      >
        AND
      </option>
      <option
        value="OR"
      >
        OR
      </option>
      <option
        value="NOT"
      >
        NOT
      </option>
    </select>
    <div
      className="dropdown-container"
    >
      <button
        className="button dropdown"
        onClick={[Function]}
        type="button"
      >
        Select
      </button>
      <div
        className="dropdown-menu"
      >
        <div
          className="dropdown-container"
        >
          <form
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Search for field"
              type="text"
              value=""
            />
          </form>
          <div
            className="autocomplete-menu"
          >
            <div
              className="dropdown-menu__panel"
            />
          </div>
        </div>
        <div
          className="dropdown-menu__panel"
        >
          <ul
            className=""
          />
        </div>
      </div>
    </div>
    <div
      className="advanced-search__inputs"
    >
      <label
        htmlFor="input_All"
      >
        Any
        <input
          id="input_All"
          onChange={[Function]}
          placeholder="a4_human, P05067, cdc7 human"
          type="text"
        />
      </label>
    </div>
    <button
      className="button-remove"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
  <div
    className="advanced-search__field"
  >
    <select
      className="advanced-search__logic"
      onChange={[Function]}
      value="AND"
    >
      <option
        value="AND"
      >
        AND
      </option>
      <option
        value="OR"
      >
        OR
      </option>
      <option
        value="NOT"
      >
        NOT
      </option>
    </select>
    <div
      className="dropdown-container"
    >
      <button
        className="button dropdown"
        onClick={[Function]}
        type="button"
      >
        Select
      </button>
      <div
        className="dropdown-menu"
      >
        <div
          className="dropdown-container"
        >
          <form
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Search for field"
              type="text"
              value=""
            />
          </form>
          <div
            className="autocomplete-menu"
          >
            <div
              className="dropdown-menu__panel"
            />
          </div>
        </div>
        <div
          className="dropdown-menu__panel"
        >
          <ul
            className=""
          />
        </div>
      </div>
    </div>
    <div
      className="advanced-search__inputs"
    >
      <label
        htmlFor="input_All"
      >
        Any
        <input
          id="input_All"
          onChange={[Function]}
          placeholder="a4_human, P05067, cdc7 human"
          type="text"
        />
      </label>
    </div>
    <button
      className="button-remove"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
  <hr />
  <div>
    <button
      className="button"
      id="add-field"
      onClick={[Function]}
      type="button"
    >
      Add Field
    </button>
    <button
      className="button"
      id="submit-query"
      onClick={[Function]}
      type="button"
    >
      Search
    </button>
  </div>
</div>
`;
